src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs

```diff
@@ -1,141 +1,80 @@
-        {
-            this._processEventSemaphoreSlim[descriptorName].Release();
-            await this._eventProcessedTaskCompletionSource.Task.ConfigureAwait(false);
-        }
-
-        this._eventProcessedTaskCompletionSource = null;
-    }
-
-    public override void Dispose()
-    {
-        if (this._processEventSemaphoreSlim is not null)
-        {
-            foreach (var kvp in this._processEventSemaphoreSlim)
-            {
-                kvp.Value.Dispose();
-            }
-        }
-
-        base.Dispose();
-    }
-
     private class EventGridSubscriptionEventPuller
     {
         private const int OutputChannelSize = 5000;
         private const int MaxEventRequestSize = 100;
 
         private readonly IServiceScopeFactory _serviceScopeFactory;
         private readonly ILogger<EventPullerService> _logger;
-        private readonly EventPullerService _pullerService;
-        private readonly string _descriptorName;
         private readonly EventGridTopicSubscription _eventGridTopicSubscription;
 
-        private readonly Channel<string> _acknowledgeEventChannel = Channel.CreateBounded<string>(OutputChannelSize);
-        private readonly Channel<string> _releaseEventChannel = Channel.CreateBounded<string>(OutputChannelSize);
-        private readonly Channel<string> _rejectEventChannel = Channel.CreateBounded<string>(OutputChannelSize);
+        private readonly Dictionary<int, TimeSpan>? _retryDelays;
+        private readonly TimeSpan? _defaultDelay;
+
+        private readonly Channel<EventBundle> _acknowledgeEventChannel = Channel.CreateBounded<EventBundle>(OutputChannelSize);
+        private readonly Channel<EventBundle> _releaseEventChannel = Channel.CreateBounded<EventBundle>(OutputChannelSize);
+        private readonly Channel<EventBundle> _rejectEventChannel = Channel.CreateBounded<EventBundle>(OutputChannelSize);
 
         private int _handlersInProgress;
         private TaskCompletionSource _taskCompletionSource = new();
 
         public EventGridSubscriptionEventPuller(
-            EventPullerService pullerService,
-            string descriptorName,
             EventGridTopicSubscription eventGridTopicSubscription,
             IServiceScopeFactory serviceScopeFactory,
             ILogger<EventPullerService> logger)
         {
             this._serviceScopeFactory = serviceScopeFactory;
             this._logger = logger;
-            this._pullerService = pullerService;
-            this._descriptorName = descriptorName;
             this._eventGridTopicSubscription = eventGridTopicSubscription;
+
+            if (eventGridTopicSubscription.RetryDelays != null)
+            {
+                this._retryDelays = eventGridTopicSubscription.RetryDelays
+                    .Select((x, index) => new { DeliveryCount = index + 1, Delay = x })
+                    .ToDictionary(x => x.DeliveryCount, x => x.Delay);
+
+                this._defaultDelay = eventGridTopicSubscription.RetryDelays.Last();
```

Comment:
What if `eventGridTopicSubscription.RetryDelays` is empty?

Reply:
Good point, I added a safety check for that